CODING_STANDARDS.md

```diff
@@ -0,0 +1,164 @@
+
+For example:
+
+```cpp
+while (x == y)
+{
+  something();
+  somethingelse();
+}
+finalthing();
+```
+
+### File names
+
+File names should be lower case with underscore word separators. C file names use the `.c` extension, C++ files use the `.cc` extension and header files use the `.h` extension.
+
+### Language standards
+
+For pure-C files we use C99 and for C++ we use C++11. The code need to be able to compile on multiple platforms using different compilers (for example: Windows / Linux, x86_64 / ARM).
+
+### Line lengths
+
+Lines should be no more than 80 characters.
+
+### Comments
+
+Single line / inline code comments can use the double slash (`//`) style of coding, whereas multi-line code comments should use `/*` as a start and `*/` at the end, with the text indented by 2 spaces, for example:
+
+```cpp
+/*
+  This is a multi-line code comment.
+  It has an indentation of two spaces.
+*/
+```
+
+### Variables classes, and functions
+
+Variables and functions should be descriptive and in "snake case", for example:
+
+```cpp
+void my_function(uint16 variable_name)
+{
+```
+
+Class names should also be "snake case" but should start with an upper-case character. Such as this:
+
+```cpp
+class Buffered_logs
+{
+```
+
+### Constant integers
+
+Constant integers that are used to define elements such as buffer sizes should defined rather than used directly. This is because the integer could change and uses of it could be missed. For example:
+
+```cpp
+char *buffer= my_malloc(PSI_INSTRUMENT_ME, 1024, MYF(MY_WME));
+
+snprint(buffer, 1024, "%d: %s", integer, text);
+```
+
+Could become:
+
+```cpp
+const int buffer_size= 1024;
+char *buffer= my_malloc(PSI_INSTRUMENT_ME, buffer_size, MYF(MY_WME));
+
+snprint(buffer, buffer_size, "%d: %s", integer, text);
+```
+
+Alternatively the integer can be defined using an `enum` or `#define`.
+
+### Spacing
+
+#### Whitespace
+
+* Lines should not have any trailing whitespace.
+* There should not be any trailing blank lines at the end of a file.
+* Line endings are POSIX style (`\n`).
+* Two spaces for each indentation level, not tabs.
+
+#### Pointers
+
+The `*` of a pointer should be on the side of the variable name such as:
+
+```cpp
+void my_function(THD *thd)
+{
+```
+
+As yet there is no standard as to whether the `*` in a casting should have a space or not. Both of these are valid:
+
+```cpp
+name= (const char*) db_name;
+name= (const char *) db_name;
+```
+
+#### Equals
+
+When setting a variable there should be no space on the left hand side and one space on the right hand side. For example:
+
+```cpp
+number= 123;
+```
+
+#### Function variables
+
+There should be a space after each comma in a definition and usage of a function. For example:
+
+```cpp
+my_function(thd, db_name);
+```
+
+### Types
+
+MariaDB has a basic type system defined in `my_global.h`, in general:
+
+* `unsigned char` -> `uchar`
+* `unsigned int` -> `uint`
+* `unsigned short` -> `ushort`
+* `unsigned long` -> `ulong`
```

Comment:
`long` and `unsigned long` are not portable. `ssize_t` and `size_t` should be preferred. While most 64-bit ABIs that we target are LP64, Microsoft Windows follows LLP64, where `sizeof(long) == 4` and `sizeof(size_t) == 8`.
I think that we should explicitly discourage the use of `long` and `ulong`.

https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

Reply:
@dr-m Agreed, good call! In fact "char" even has portability issues as it can be signed or unsigned depending on platform :)

Reply:
@dr-m @LinuxJedi this highlights something that should be its own issue: documentation lacks real guidance on when to use which types. Depending on the usage `size_t` can even be a very poor choice, for example in serialization one might choose `uint64_t` (or perhaps even `uint_least64_t`) in order to exchange data more easily between 32bit and 64bit platforms.

Regardless, rather than expand the scope of this change, I'd suggest that highlighting that preferred types are defined in `include/my_global.h`.

Reply:
> @dr-m Agreed, good call! In fact "char" even has portability issues as it can be signed or unsigned depending on platform :)

For certain, and in real life others have been hit by that one, e.g.:
https://github.com/skarnet/skalibs/commit/52b7288713ecb189fdd473f9dfc7f9a8c4fb1ec2

Reply:
I think that  ulong is not good to use in any new API would be good enough. It is good for exactly nothing, neither serialization (fixed types are good), nor buffer sizes (size_t is good), nor file lengths or offsets (64bit intergers are good). ulong either brings warnings, and casts to silence the warnings, or it can be replaced with uint.

Reply:
Why not use the [stdint.h types](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdint.h.html) when a specific size is needed?

- `uint32_t`, `int8_t` when you need a specific size (e.g. for serialization)
- `uintptr_t`, `intptr_t` when you need an integer of a size that can be round-trip casted/type-punned with a `void *`

Lots of FLOSS C/C++ projects make extensive use of these nowadays.

Reply:
> Why not use the [stdint.h types](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdint.h.html) when a specific size is needed?
> 
>     * `uint32_t`, `int8_t` when you need a specific size (e.g. for serialization)
> 
>     * `uintptr_t`, `intptr_t` when you need an integer of a size that can be round-trip casted/type-punned with a `void *`
> 
> 
> Lots of FLOSS C/C++ projects make extensive use of these nowadays.

There used to be portability issues with this a long time ago. I don't think it is so much of an issue any more for POSIX and Windows at least (definitely still a problem on other platforms that we don't really support). It could be revisited, but way out of scope for this PR. This is more of s descriptive PR of what is happening right now. Debates on changing the source code's style can happen later :)